core/src/commons/Data.cpp

```diff
@@ -176,66 +174,32 @@ void Data::get_all_values(std::vector<double>& all_values,
   std::iota(index.begin(), index.end(), 0);
   // sort index based on the split values (argsort)
   // the NaN comparison places all NaNs at the beginning
+  // stable sort is needed for consistent element ordering cross platform,
+  // otherwise the resulting sums used in the splitting rules may compound rounding error
+  // differently and produce different splits.
   std::stable_sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs) {
-    return all_values[lhs] < all_values[rhs] || (std::isnan(all_values[lhs]) && !std::isnan(all_values[rhs]));
+    double left_val = all_values[lhs];
```

Comment:
What's the motivation for the change here, and for the related changes in the splitting rules? Is it just a straight refactor? If so it'd be best to leave it out of this PR, since it doesn't seem related to removing the big Q splitting rule. (To me the refactor isn't an obvious improvement either, as it adds complexity into the splitting rules around retrieving the right 'best split value').

Reply:
This simple change was just meant to remove two additional (potentially) cache expensive retrievals of all_values[.], that was probably unnecessary the compiler should optimize that away? I can restore it to the original.

The change in the splitting rule comes from removing the use of `possible_split_values` ( skipping an `erase` call). You mentioned it here https://github.com/grf-labs/grf/pull/592#discussion_r360650548 and I thought it made sense to do here, should I restore it? (I can just revert commit b4ab0af)

Reply:
My original comment was unclear -- by 'the change here' I meant the changes to the method as a whole. I think it'd make sense to revert the changes to the method (including commit b4ab0af as you mention), since it's separate from the goal of the PR. And I'm not sure it helps simplify or clarify the logic now that we have to make changes to the splitting rules.

Reply:
Ok, it has been reverted